File: tests/wasm/test_ehlers_pma.js
/**
 * WASM binding tests for Ehlers PMA indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/my_project.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('Ehlers PMA accuracy', () => {
    
    const close = new Float64Array(testData.close);
    
    
    const result = wasm.ehlers_pma(close);
    
    
    const predict = result.values.slice(0, close.length);
    const trigger = result.values.slice(close.length);
    
    assert.strictEqual(predict.length, close.length);
    assert.strictEqual(trigger.length, close.length);
    
    
    const expectedPredictLastFive = [
        59161.97066327,
        59240.51785714,
        59260.29846939,
        59225.19005102,
        59192.78443878
    ];
    const expectedTriggerLastFive = [
        59020.56403061,
        59141.96938776,
        59214.56709184,
        59232.46619898,
        59220.78227041
    ];
    
    
    assertArrayClose(
        predict.slice(-5),
        expectedPredictLastFive,
        1e-6,
        "Ehlers PMA predict last 5 values mismatch"
    );
    
    assertArrayClose(
        trigger.slice(-5),
        expectedTriggerLastFive,
        1e-6,
        "Ehlers PMA trigger last 5 values mismatch"
    );
});

test('Ehlers PMA default candles', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.ehlers_pma(close);
    
    
    assert.strictEqual(result.values.length, close.length * 2);
});

test('Ehlers PMA empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.ehlers_pma(empty);
    }, /Input data slice is empty/);
});

test('Ehlers PMA all NaN values', () => {
    
    const allNan = new Float64Array(20).fill(NaN);
    
    assert.throws(() => {
        wasm.ehlers_pma(allNan);
    }, /All values are NaN/);
});

test('Ehlers PMA insufficient data', () => {
    
    const data = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0]);
    
    assert.throws(() => {
        wasm.ehlers_pma(data);
    }, /Not enough valid data/);
});

test('Ehlers PMA very small dataset', () => {
    
    const singlePoint = new Float64Array([42.0]);
    
    assert.throws(() => {
        wasm.ehlers_pma(singlePoint);
    }, /Not enough valid data/);
    
    const twoPoints = new Float64Array([42.0, 43.0]);
    
    assert.throws(() => {
        wasm.ehlers_pma(twoPoints);
    }, /Not enough valid data/);
});

test('Ehlers PMA NaN handling', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.ehlers_pma(close);
    const predict = result.values.slice(0, close.length);
    const trigger = result.values.slice(close.length);
    
    assert.strictEqual(predict.length, close.length);
    assert.strictEqual(trigger.length, close.length);
    
    
    if (predict.length > 20) {
        for (let i = 20; i < predict.length; i++) {
            assert(!isNaN(predict[i]), `Found unexpected NaN in predict at index ${i}`);
        }
        for (let i = 20; i < trigger.length; i++) {
            assert(!isNaN(trigger[i]), `Found unexpected NaN in trigger at index ${i}`);
        }
    }
    
    
    
    for (let i = 0; i < 13; i++) {
        assert(isNaN(predict[i]), `Expected NaN in predict warmup at index ${i}`);
    }
    
    for (let i = 0; i < 16; i++) {
        assert(isNaN(trigger[i]), `Expected NaN in trigger warmup at index ${i}`);
    }
});

test('Ehlers PMA with minimum required data', () => {
    
    const data = new Float64Array(Array.from({length: 14}, (_, i) => 100.0 + i));
    
    const result = wasm.ehlers_pma(data);
    assert.strictEqual(result.values.length, data.length * 2);
    
    const predict = result.values.slice(0, data.length);
    const trigger = result.values.slice(data.length);
    
    
    const validPredict = [];
    for (let i = 0; i < predict.length; i++) {
        if (!isNaN(predict[i])) {
            validPredict.push(predict[i]);
        }
    }
    assert(validPredict.length > 0, 'Should have at least one valid predict value');
});

test('Ehlers PMA with constant values', () => {
    
    const data = new Float64Array(50).fill(100.0);
    
    const result = wasm.ehlers_pma(data);
    const predict = result.values.slice(0, data.length);
    
    
    let validIdx = -1;
    for (let i = 0; i < predict.length; i++) {
        if (!isNaN(predict[i])) {
            validIdx = i;
            break;
        }
    }
    
    assert(validIdx !== -1, 'Should have valid predict values');
    
    
    assertClose(predict[validIdx], 100.0, 1e-9, 'Predict should equal constant value');
});

test('Ehlers PMA crossover detection', () => {
    
    const close = new Float64Array(testData.close.slice(0, 200));
    
    const result = wasm.ehlers_pma(close);
    const predict = result.values.slice(0, close.length);
    const trigger = result.values.slice(close.length);
    
    
    const crossovers = [];
    for (let i = 17; i < predict.length - 1; i++) {
        if (!isNaN(predict[i]) && !isNaN(trigger[i]) && 
            !isNaN(predict[i-1]) && !isNaN(trigger[i-1])) {
            if (predict[i-1] <= trigger[i-1] && predict[i] > trigger[i]) {
                crossovers.push({type: 'bullish', index: i});
            } else if (predict[i-1] >= trigger[i-1] && predict[i] < trigger[i]) {
                crossovers.push({type: 'bearish', index: i});
            }
        }
    }
    
    
    assert(crossovers.length > 0, 'Should detect at least one crossover in real data');
});

test('Ehlers PMA with NaN in middle of data', () => {
    
    
    
    const data = new Float64Array(30);
    for (let i = 0; i < data.length; i++) {
        data[i] = 100.0 + i;
    }
    
    
    data[7] = NaN;
    
    const result = wasm.ehlers_pma(data);
    assert.strictEqual(result.values.length, data.length * 2);
    
    const predict = result.values.slice(0, data.length);
    
    
    
    
    
    
    for (let i = 0; i < 15; i++) {
        assert(isNaN(predict[i]), `Expected NaN at index ${i} during warmup/affected period`);
    }
    
    
    
    let hasValidAfterNaN = false;
    for (let i = 15; i < predict.length; i++) {
        if (!isNaN(predict[i])) {
            hasValidAfterNaN = true;
            break;
        }
    }
    assert(hasValidAfterNaN, 'Should have valid values after NaN falls out of 7-period window');
});

File: tests/python/test_ehlers_pma.py
"""
Python binding tests for Ehlers PMA indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import my_project as rb
from test_utils import load_test_data
import os

class TestEhlersPma:
    @pytest.fixture(scope='class')
    def test_data(self):
        """Load test data for Ehlers PMA indicator."""
        return load_test_data()

    def test_ehlers_pma_accuracy(self, test_data):
        """Test Ehlers PMA matches expected values from Rust tests - mirrors check_ehlers_pma_accuracy"""
        
        close = test_data['close']
        predict, trigger = rb.ehlers_pma(close)
        
        
        expected_predict_last_five = [
            59161.97066327,
            59240.51785714,
            59260.29846939,
            59225.19005102,
            59192.78443878,
        ]
        expected_trigger_last_five = [
            59020.56403061,
            59141.96938776,
            59214.56709184,
            59232.46619898,
            59220.78227041,
        ]
        
        assert len(predict) == len(close)
        assert len(trigger) == len(close)
        
        
        np.testing.assert_allclose(
            predict[-5:],
            expected_predict_last_five,
            rtol=1e-6,
            err_msg="Ehlers PMA predict last 5 values mismatch"
        )
        
        np.testing.assert_allclose(
            trigger[-5:],
            expected_trigger_last_five,
            rtol=1e-6,
            err_msg="Ehlers PMA trigger last 5 values mismatch"
        )

    def test_ehlers_pma_default_candles(self, test_data):
        """Test Ehlers PMA with default parameters - mirrors check_ehlers_pma_default_candles"""
        close = test_data['close']
        
        
        predict, trigger = rb.ehlers_pma(close)
        assert len(predict) == len(close)
        assert len(trigger) == len(close)

    def test_ehlers_pma_empty_input(self):
        """Test Ehlers PMA fails with empty input - mirrors check_ehlers_pma_empty_input"""
        empty = np.array([])
        with pytest.raises(ValueError, match="Input data slice is empty"):
            rb.ehlers_pma(empty)

    def test_ehlers_pma_all_nan(self):
        """Test Ehlers PMA fails with all NaN values - mirrors check_ehlers_pma_all_nan"""
        all_nan = np.full(20, np.nan)
        with pytest.raises(ValueError, match="All values are NaN"):
            rb.ehlers_pma(all_nan)

    def test_ehlers_pma_insufficient_data(self):
        """Test Ehlers PMA fails with insufficient data - mirrors check_ehlers_pma_insufficient_data"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
        with pytest.raises(ValueError, match="Not enough valid data"):
            rb.ehlers_pma(data)
            
    def test_ehlers_pma_very_small_dataset(self):
        """Test Ehlers PMA fails with very small dataset - mirrors check_ehlers_pma_very_small_dataset"""
        single_point = np.array([42.0])
        with pytest.raises(ValueError, match="Not enough valid data"):
            rb.ehlers_pma(single_point)
            
        two_points = np.array([42.0, 43.0])
        with pytest.raises(ValueError, match="Not enough valid data"):
            rb.ehlers_pma(two_points)

    def test_ehlers_pma_nan_handling(self, test_data):
        """Test Ehlers PMA handles NaN values correctly - mirrors check_ehlers_pma_nan_handling"""
        close = test_data['close']
        
        predict, trigger = rb.ehlers_pma(close)
        assert len(predict) == len(close)
        assert len(trigger) == len(close)
        
        
        if len(predict) > 20:
            
            assert not np.any(np.isnan(predict[20:])), "Found unexpected NaN in predict after warmup"
            
            assert not np.any(np.isnan(trigger[20:])), "Found unexpected NaN in trigger after warmup"
        
        
        assert np.all(np.isnan(predict[:13])), "Expected NaN in predict warmup period (first 13)"
        
        assert np.all(np.isnan(trigger[:16])), "Expected NaN in trigger warmup period (first 16)"
    
    def test_ehlers_pma_streaming(self, test_data):
        """Test Ehlers PMA streaming matches batch calculation - mirrors check_ehlers_pma_streaming"""
        close = test_data['close'][:100]  
        
        
        batch_predict, batch_trigger = rb.ehlers_pma(close)
        
        
        stream = rb.EhlersPmaStream()
        stream_values = []
        
        for price in close:
            result = stream.update(price)
            stream_values.append(result)
        
        
        stream_predict = []
        stream_trigger = []
        for result in stream_values:
            if result is None:
                stream_predict.append(np.nan)
                stream_trigger.append(np.nan)
            else:
                predict, trigger = result
                stream_predict.append(predict)
                stream_trigger.append(trigger)
        
        stream_predict = np.array(stream_predict)
        stream_trigger = np.array(stream_trigger)
        
        
        
        for i in range(17, len(close)):  
            if not np.isnan(batch_predict[i]) and not np.isnan(stream_predict[i]):
                np.testing.assert_allclose(
                    batch_predict[i], 
                    stream_predict[i], 
                    rtol=1e-9, 
                    err_msg=f"Predict streaming mismatch at index {i}"
                )
            if not np.isnan(batch_trigger[i]) and not np.isnan(stream_trigger[i]):
                np.testing.assert_allclose(
                    batch_trigger[i], 
                    stream_trigger[i], 
                    rtol=1e-9,
                    err_msg=f"Trigger streaming mismatch at index {i}"
                )

    def test_ehlers_pma_batch(self, test_data):
        """Test Ehlers PMA batch processing - mirrors batch functionality"""
        close = test_data['close']
        
        
        result = rb.ehlers_pma_batch(close)
        
        assert 'values' in result
        assert 'lines' in result  
        
        
        assert result['values'].shape[0] == 2  
        assert result['values'].shape[1] == len(close)
        
        
        predict_row = result['values'][0]
        trigger_row = result['values'][1]
        
        
        expected_predict_last_five = [
            59161.97066327,
            59240.51785714,
            59260.29846939,
            59225.19005102,
            59192.78443878,
        ]
        expected_trigger_last_five = [
            59020.56403061,
            59141.96938776,
            59214.56709184,
            59232.46619898,
            59220.78227041,
        ]
        
        
        np.testing.assert_allclose(
            predict_row[-5:],
            expected_predict_last_five,
            rtol=1e-6,
            err_msg="Ehlers PMA batch predict mismatch"
        )
        
        np.testing.assert_allclose(
            trigger_row[-5:],
            expected_trigger_last_five,
            rtol=1e-6,
            err_msg="Ehlers PMA batch trigger mismatch"
        )


if __name__ == '__main__':
    pytest.main([__file__, '-v'])
